### PR TITLE
azure-cli@2.0 0.2.10 (new formula)

### PR DESCRIPTION
**Pull request by antoineco**
---

**Moved to a third-party Tap:** [antoineco/azure-cli](https://github.com/antoineco/homebrew-azure-cli), **until upstream fixes the blocker described below.**

---

[Azure CLI 2.0](https://github.com/Azure/azure-cli) installed from the [packaged releases](https://github.com/Azure/azure-cli/blob/master/packaged_releases/HISTORY.md) distributed by Microsoft.

**Info**: https://docs.microsoft.com/en-us/cli/azure/overview
**Release notes**: https://docs.microsoft.com/en-us/cli/azure/release-notes-azure-cli

----

<details>
  <summary>GitHub PR template</summary>

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

</details>

### CODE REVIEW

**Comment by fxcoudert** [comment]
We currently ship `azure-cli: stable 0.10.12`. What is the relationship between the two? Is 2.0 a version name of just part of the marketing? If so, this formula should be `azure-cli-2.0` with no `@` and no alias.

**Comment by antoineco** [comment]
It's a complete rewrite from node to Python, this new CLI also contains new features.

**Comment by antoineco** [comment]
@fxcoudert the [*Get Started with Azure*](https://docs.microsoft.com/en-us/azure/index#pivot=services&panel=get-started) page actually refers to the 2.0 CLI as the standard command line tool for Azure already.

Should we rather raname the old `azure-cli` formula to `azure-cli@0.10` and use this new formula as the standard `azure-cli`?

**Comment by antoineco** [comment]
@ilovezfs any problem I can help with in order to get this merged?

**Comment by antoineco** [comment]
I went ahead and renamed `azure-cli` to `azure-cli@0.10`. I'd appreciate some feedback on that.

**Comment by dunn** [comment]
Why do we need both formulae? Why not just update the existing one to 2.0?

**Comment by antoineco** [comment]
@dunn I'm all for replacing the current formula, as long as nobody wants to keep using the  "old" CLI. However brew will complain that the version number can not decrease (0.10 -> 0.2).

**Comment by xyziemba** [comment]
Azure CLI 2.0 doesn't support managing "classic" Azure Service Manager deployments. The existing tool does. Replacing the formula breaks people who use the existing tool to manage older deployments. Also, the existing formula isn't abandonware. There's a new version every month or so. Both Azure CLI 2.0 and Azure CLI 1.0 appear to be under active development.

IMO, Azure CLI 2.0 should live in a new formula. It's a different product, and it has a different codebase. Perhaps that formula could be named something like `az`, `azure-az-cli`, `az-cli`, or `azure-cli-2.0`.

**Comment by ilovezfs** [comment]
@antoineco do you yourself need 1.0?

**Comment by ilovezfs** [comment]
@xyziemba do you yourself need 1.0?

**Comment by antoineco** [comment]
@ilovezfs I don't, but it doesn't matter what **I** use. As @xyziemba said these are 2 different codebases, which 2 different set of features: 1 "legacy", 1 "future".

**Comment by xyziemba** [comment]
@ilovezfs Yes, I do. That's why I update this formula every time there's a new release of 1.0. 😉

**Comment by ilovezfs** [comment]
OK, so how does this sound? I'm thinking we should upgrade the azure-cli formula to be 2.0, and avoid the "version number cannot decrease" issue by adding `version_scheme 1` to the formula. Then separately we can add a new formula `azure@1.0` or `azure@1`.

**Comment by antoineco** [comment]
@ilovezfs let me push some separate commits for that then. Thanks for your feedback, much appreciated!

**Comment by antoineco** [comment]
@ilovezfs I'm facing a minor issue while renaming the 2.0 formula:
```
❯ brew audit --strict azure-cli              
azure-cli:
  * Formula has other versions so create a versioned alias:
      cd /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Aliases
      ln -s ../Formula/azure-cli.rb azure-cli@0.2
```
Is it possible to have a "2.0" formula following an [alternative version scheme](https://github.com/Azure/azure-cli/blob/master/packaged_releases/HISTORY.md)?

**Comment by ilovezfs** [comment]
@MikeMcQuaid any thoughts on how to handle upstream's messy versioning here? They haven't done us any favors in terms of mapping this onto the `@` stuff by having their 1.0 versions not start with "1.0" and their 2.0 versions not start with "2.0"

**Comment by antoineco** [comment]
Could we make it ugly and rename azure-cli to azure-cli-1 for example?

**Comment by MikeMcQuaid** [comment]
@ilovezfs I guess we just have to abuse `version_scheme`. Alternatively, we make up our own `version`s for these based on actually using semantic versioning.

File: Formula/azure-cli@2.0.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  head "https://github.com/Azure/azure-cli.git"
+
+  depends_on :python
+
+  def install
+    virtualenv_create(libexec)
+    bin_dir = libexec/"bin"
+
+    # Install CLI components and dependencies
+    Dir.mktmpdir do |tmp_dir|
```
Comment (MikeMcQuaid) [inline Formula/azure-cli@2.0.rb]: Why can this not be done in the existing temporary directory Homebrew already runs `install` in?

File: Formula/azure-cli@2.0.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      components = [
+        "src/azure-cli",
+        "src/azure-cli-core",
+        "src/azure-cli-nspkg",
+        "src/azure-cli-command_modules-nspkg",
+      ] + Pathname.glob("src/command_modules/azure-cli-*/")
```
Comment (MikeMcQuaid) [inline Formula/azure-cli@2.0.rb]: What does this glob return?

File: Formula/azure-cli@2.0.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+      # Create wheel distribution of included components
+      components.each do |c|
+        Dir.chdir(c) { system bin_dir/"python", "setup.py", "bdist_wheel", "-d", tmp_dir }
```
Comment (MikeMcQuaid) [inline Formula/azure-cli@2.0.rb]: Build an array of `Pathname` and use `Pathname#cd` instead.

File: Formula/azure-cli@2.0.rb [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      end
+
+      # Install CLI from wheel distributions
+      system bin_dir/"pip", "install", "azure-cli", "-f", tmp_dir
+    end
+
+    # Generate executable
+    File.write("#{bin_dir}/az", "#!/usr/bin/env bash\n#{bin_dir}/python -m azure.cli \"$@\"")
```
Comment (MikeMcQuaid) [inline Formula/azure-cli@2.0.rb]: We should aim to get this script upstream. Also, use `EOS.undent` rather than embedding newlines.

File: Formula/azure-cli@2.0.rb [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    bin.install_symlink bin_dir/"az"
+
+    # Install bash completion
+    bash_completion.install "az.completion" => "az"
+  end
+
+  test do
+    run_output = shell_output("#{bin}/az")
+    assert_match "Azure CLI", run_output
```
Comment (MikeMcQuaid) [inline Formula/azure-cli@2.0.rb]: I don't think we need this test along with `--version`, let's pick one.

File: Formula/azure-cli@2.0.rb [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+    version_output = shell_output("#{bin}/az --version")
+    assert_match "azure-cli", version_output
+
+    assert_match "-F _python_argcomplete az",
+      shell_output("source #{bash_completion}/az && complete -p az")
```
Comment (MikeMcQuaid) [inline Formula/azure-cli@2.0.rb]: This seems pretty implementation specific and can be likely removed.

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]

**Comment by ilovezfs** [comment]
So in talking with @MikeMcQuaid about this privately, I think we'll want to use our own version numbers here and map

azure-cli 2.0 upstream version 0.2.9 to Homebrew version

```
version "2.0.2.9"
```

and azure-cli 1.0 upstream version 0.10.13 to Homebrew version
```
version "1.0.10.13"
```

@antoineco Does that make sense?

File: Formula/azure-cli@2.0.rb (hunk 1 above)
Comment (antoineco) [inline Formula/azure-cli@2.0.rb]: How do I reference that dir?

Comment (ilovezfs) [inline Formula/azure-cli@2.0.rb]: `buildpath`

File: Formula/azure-cli@2.0.rb (hunk 2 above)
Comment (antoineco) [inline Formula/azure-cli@2.0.rb]: The list of modules in [this directory](https://github.com/Azure/azure-cli/tree/master/src/command_modules).

Comment (MikeMcQuaid) [inline Formula/azure-cli@2.0.rb]: Thanks.

**Comment by antoineco** [comment]
@ilovezfs totally, I'll push these changes shortly.

**Comment by ilovezfs** [comment]
@antoineco thanks!

File: Formula/azure-cli@2.0.rb (hunk 4 above)
Comment (antoineco) [inline Formula/azure-cli@2.0.rb]: I agree, the best I could find upstream was [that script](https://github.com/Azure/azure-cli/blob/master/src/azure-cli/az). Let me see if that would work.

Comment (antoineco) [inline Formula/azure-cli@2.0.rb]: Doesn't work unfortunately, it sets `PYTHONPATH=/usr/local/opt/python/bin/python2.7` which is not what we want.

File: Formula/azure-cli@2.0.rb (hunk 3 above)
Comment (antoineco) [inline Formula/azure-cli@2.0.rb]: Not sure which method you're referring to? I only know `Dir::chdir` and can't find `Pathname#cd` in the Ruby doc.

File: Formula/azure-cli@2.0.rb (hunk 4 above)
Comment (MikeMcQuaid) [inline Formula/azure-cli@2.0.rb]: Why is that not what we want?

File: Formula/azure-cli@2.0.rb (hunk 3 above)
Comment (MikeMcQuaid) [inline Formula/azure-cli@2.0.rb]: It's a Homebrew method. You can use `c.cd do` as a block instead of `Dir.chdir`

File: Formula/azure-cli@2.0.rb (hunk 4 above)
Comment (antoineco) [inline Formula/azure-cli@2.0.rb]: Because this Formula is installed within a Python virtualenv, so we must use the `python` binary matching that virtualenv. The "main" Python doesn't know about the packages installed by this Formula.

File: Formula/azure-cli@2.0.rb (hunk 3 above)
Comment (antoineco) [inline Formula/azure-cli@2.0.rb]: Found it inside [`pathname.rb`](https://github.com/Homebrew/brew/blob/45a7730f61ab24a0e5e3bc00d757a4521c2e1c1d/Library/Homebrew/extend/pathname.rb#L336-L338), cheers

**Comment by antoineco** [comment]
@MikeMcQuaid @ilovezfs friendly ping because GitHub doesn't notify on new/edited commits.

File: Formula/azure-cli.rb [hunk 7]
```diff
@@ -1,36 +1,46 @@
...
-  depends_on "node"
-  depends_on :python => :build
+  desc "Microsoft Azure CLI 2.0"
+  homepage "https://docs.microsoft.com/en-us/cli/azure/overview"
+  url "https://azurecliprod.blob.core.windows.net/releases/azure-cli_packaged_0.2.9.tar.gz"
+  version "2.0.2.9"
+  sha256 "6f949162d6bbf6848177cc009d60520c12cf2c1dfc068e9e5bbeac0ff6e55e4f"
+  head "https://github.com/Azure/azure-cli.git"
+
+  depends_on :python
```
Comment (ilovezfs) [inline Formula/azure-cli.rb]: would prefer
```
depends_on :python if MacOS.version <= :snow_leopard
```
assuming it works.

File: Formula/azure-cli.rb [hunk 8]
```diff
@@ -1,36 +1,46 @@
...
+    components = [
+      "src/azure-cli",
+      "src/azure-cli-core",
+      "src/azure-cli-nspkg",
+      "src/azure-cli-command_modules-nspkg",
+    ].map { |x| Pathname.new(x) } + Pathname.glob("src/command_modules/azure-cli-*/")
+
+    # Create wheel distribution of included components
+    components.each do |c|
+      c.cd { system bin_dir/"python", "setup.py", "bdist_wheel", "-d", buildpath/"wheel" }
```
Comment (ilovezfs) [inline Formula/azure-cli.rb]: does `venv.pip_install` not work for these? (see certbot)

File: Formula/azure-cli.rb [hunk 9]
```diff
@@ -1,36 +1,46 @@
...
   def install
-    rm_rf "bin/windows"
-    system "npm", "install", *Language::Node.std_npm_install_args(libexec)
-    bin.install_symlink Dir["#{libexec}/bin/*"]
-    (bash_completion/"azure").write Utils.popen_read("#{bin}/azure --completion")
+    virtualenv_create(libexec)
+    bin_dir = libexec/"bin"
+
+    components = [
+      "src/azure-cli",
```
Comment (ilovezfs) [inline Formula/azure-cli.rb]: should be able to use buildpath/"foo" for these if you need Pathname objects

File: Formula/azure-cli.rb [hunk 10]
```diff
@@ -1,36 +1,46 @@
...
+      "src/azure-cli-nspkg",
+      "src/azure-cli-command_modules-nspkg",
+    ].map { |x| Pathname.new(x) } + Pathname.glob("src/command_modules/azure-cli-*/")
+
+    # Create wheel distribution of included components
+    components.each do |c|
+      c.cd { system bin_dir/"python", "setup.py", "bdist_wheel", "-d", buildpath/"wheel" }
+    end
+
+    # Install CLI from wheel distributions
```
Comment (ilovezfs) [inline Formula/azure-cli.rb]: is this using wheel binaries?

File: Formula/azure-cli.rb [hunk 11]
```diff
@@ -1,36 +1,46 @@
...
+    system bin_dir/"pip", "install", "azure-cli", "-f", buildpath/"wheel"
```
Comment (ilovezfs) [inline Formula/azure-cli.rb]: venv.pip_install

File: Formula/azure-cli.rb [hunk 12]
```diff
@@ -1,36 +1,46 @@
...
+
+    # Generate executable
```
Comment (ilovezfs) [inline Formula/azure-cli.rb]: @tdsmith would there be a more idiomatic way of doing this?

File: Formula/azure-cli@1.0.rb [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+  bottle do
+    cellar :any_skip_relocation
+    sha256 "0849685ed65f1bfa991c3ffbfaa281f9051d33b626b79b12b1f5341f5925976d" => :sierra
+    sha256 "6bcf4ea33dad69d8c703724e09468e08c98d053456fb41a94f6b5877d9a8c243" => :el_capitan
+    sha256 "16184e20ae089b157c2e513c1b194ace66d1f46da5eb2869fb9261b1f71aaf1f" => :yosemite
+  end
+
+  depends_on "node"
+  depends_on :python => :build
```
Comment (ilovezfs) [inline Formula/azure-cli@1.0.rb]: this shouldn't be necessary

File: Formula/azure-cli@1.0.rb [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+class AzureCliAT10 < Formula
+  desc "Official Azure CLI"
+  homepage "https://github.com/azure/azure-xplat-cli"
+  url "https://github.com/Azure/azure-xplat-cli/archive/v0.10.12-May2017.tar.gz"
+  version "1.0.10.13"
+  sha256 "60194e770b8dca0485db9d4c99f9cd432f7b43096cc5ad0353f52fd5b1b29181"
+  head "https://github.com/azure/azure-xplat-cli.git", :branch => "dev"
+
+  bottle do
```
Comment (ilovezfs) [inline Formula/azure-cli@1.0.rb]: can drop the bottle block since it's actually a new formula

File: Formula/azure-cli.rb (hunk 9 above)
Comment (antoineco) [inline Formula/azure-cli.rb]: Thanks for the tip (y)

File: Formula/azure-cli.rb (hunk 8 above)
Comment (antoineco) [inline Formula/azure-cli.rb]: Unfortunately not, the process is a bit particular here:
1. the release archive contains several modules which are built locally (in the wheel format or not)
2. `pip install` is then executed on the local "dist" dir (actually I called it `wheel` here) so that locally built packages are installed in priority over whatever version is currently available on PyPi.

File: Formula/azure-cli.rb (hunk 10 above)
Comment (antoineco) [inline Formula/azure-cli.rb]: Yes, which are present by default when the venv is created with `virtualenv_create()`.

File: Formula/azure-cli.rb (hunk 11 above)
Comment (antoineco) [inline Formula/azure-cli.rb]: I looked into that option but it did not allow me to pass `-f #{buildpath}/wheel`.

File: Formula/azure-cli@1.0.rb (hunk 13 above)
Comment (antoineco) [inline Formula/azure-cli@1.0.rb]: I just renamed that Formula, I think changes to the old azure-cli formula should be part of another PR:

**Comment by antoineco** [comment]
@ilovezfs I pushed all suggested changes (except for that `pip install` topic).

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by ilovezfs** [comment]
This will need to be rebased to resolve the conflicts due to https://github.com/Homebrew/homebrew-core/pull/14152.

**Comment by antoineco** [comment]
@ilovezfs done

**Comment by ilovezfs** [comment]
@antoineco thank you. Sorry for the hassle.

**Comment by antoineco** [comment]
Looks like azure-cli@1.0 is making the check fail:
```
Already downloaded: /Users/brew/Library/Caches/Homebrew/azure-cli@1.0-1.0.10.13.tar.gz
==> Verifying azure-cli@1.0-1.0.10.13.tar.gz checksum
tar xzf /Users/brew/Library/Caches/Homebrew/azure-cli@1.0-1.0.10.13.tar.gz
Error: No such file or directory - npm
```

**Comment by ilovezfs** [comment]
Yeah it needs to go through CI again now that the brew side of the npm 5 changes has gone through.

**Comment by antoineco** [comment]
Ok, rebased to trigger the build (good ol' ugly way).

**Comment by ilovezfs** [comment]
That'll do the trick. Thanks!

**Review by ilovezfs — APPROVED** [review]

File: Formula/azure-cli.rb (hunk 10 above)
Comment (ilovezfs) [inline Formula/azure-cli.rb]: Actually we need to build these from source. CC @tdsmith am I missing something or is this building from binaries since `--no-binary all:` isn't being passed below and it's not using our `venv.pip_install`?

**Review by ilovezfs — CHANGES_REQUESTED** [review]

**Comment by antoineco** [comment]
@ilovezfs (can't respond directly to your comment) the wheel packages *are* built from source, 2 lines above.

**Comment by ilovezfs** [comment]
@antoineco But if they have any dependencies, the absence of `"--no-binary", ":all"` means it could potentially pull in other binary wheels (unless `-f` somehow makes that impossible?).

**Comment by antoineco** [comment]
Got it, you're right, dependencies need of course to be fetched from PyPI. I ignored that flag.

**Comment by ilovezfs** [comment]
@antoineco just spoke with tdsmith privately and he should be able to share some thoughts he has for you later today.

**Comment by ilovezfs** [comment]
@antoineco you'll also want to hard reset your local branch as @JCount separated the two commits in the PR slightly differently at my request. There shouldn't be any net diff to the files.

**Comment by antoineco** [comment]
I tried running `setup.py` with `sdist` in order to avoid building binaries, but then `pip install --no-binaries` fails because some dependencies like [azure-nspkg](https://pypi.python.org/pypi/azure-nspkg/2.0.0) do not have any source distribution.

**Comment by tdsmith** [comment]
I think this is their supported installer: https://github.com/Azure/azure-cli/blob/master/scripts/curl_install_pypi/install.py

This package seems gratuitously eclectic. I'm not sure we can package it to our standards.

Just installing azure-cli from pypi and running poet over it ought to work but doesn't because they've misunderstood the ~= comparator and their packages have mutually incompatible requirements. If they also have packages without sdists for some reason, that's not going to work well with our tooling either.

**Comment by tdsmith** [comment]
It would also be a very long list of requirements (what if Python... were node)?

```
(tmp-e5e3950534343eee) tim@rocketman:tmp-e5e3950534343eee$ pip freeze | wc
     118     118    2609
```

@antoineco I'm sorry to parachute in and trash this PR after you've put so much effort into it :( I'd love to be able to ship this tool. If you're willing to help Microsoft resolve the conflicted requirements in the dependencies of `azure-cli`, I think that's the biggest blocker.

It's possible poet should accept pure-Python wheels but I don't think anyone's had a good reason to ship one of those without a sdist yet.

**Comment by antoineco** [comment]
@tdsmith you're doing your maintainer's job and keeping the standards high, I understand that.
I won't lie, I would have preferred to read that *before* investing so much time in the PR, but the work is not lost. I created a Tap that one can use to install the Formula as it is: [antoineco/azure-cli](https://github.com/antoineco/homebrew-azure-cli).

I'm done with that, don't have any more motivation to hunt the devs of `azure-cli`, but maybe @derekbekoe can do something about it if it ever becomes a priority.

**Comment by derekbekoe** [comment]
> "If you're willing to help Microsoft resolve the conflicted requirements in the dependencies of azure-cli, I think that's the biggest blocker."
@tdsmith Can you clarify this? What are the conflicted requirements?

@antoineco Thanks for starting this PR! I started one a few months ago https://github.com/Homebrew/homebrew-core/pull/8669/files but this one looks great and we'd want to see Azure CLI in homebrew.

**Comment by tdsmith** [comment]
> @tdsmith Can you clarify this? What are the conflicted requirements?

If you `pip install azure-cli`, the installed package set will have mutually incompatible requirements. This is both possible and a little difficult to observe because pip doesn't attempt to do any real dependency solving.

You can observe this in a kinda clunky way with Homebrew's tooling by creating a new virtualenv, running `pip install azure-cli homebrew-pypi-poet`, and then `poet -f azure-cli`. First it'll complain about `requests`, which you can solve with `pip install 'requests~=2.14.1'`. Rerunning poet surfaces a conflict between pins on `azure-common`; one package demands `==1.1.4` and another demands `~=1.1.5` which are incompatible specs.

Poking around the site-packages folder with `grep azure-common *dist-info/METADATA` reveals that the incompatible pin is in azure_mgmt_authorization.

**Comment by derekbekoe** [comment]
@tdsmith Ah okay. Yes we have seen these dependency mismatches as we have a CI script that checks this - https://github.com/Azure/azure-cli/blob/master/scripts/automation/tests/verify_dependencies.py#L14.

Is this a blocker for this PR? If so, can you help me understand why?

Thanks!

**Comment by tdsmith** [comment]
It's a blocker; our tooling for writing resource stanzas for python dependencies needs to know which version to install and declines to guess ("crashes") in the presence of conflicts.

**Comment by derekbekoe** [comment]
Okay thanks for the clarification.
